src/resolve/resolved-dns-query.c

```diff
@@ -1140,3 +1140,10 @@ bool dns_query_fully_confidential(DnsQuery *q) {
 
         return FLAGS_SET(q->answer_query_flags, SD_RESOLVED_CONFIDENTIAL) && !q->previous_redirect_non_confidential;
 }
+
+bool dns_query_fully_synthetic(DnsQuery *q) {
+        assert(q);
+
+        return (q->answer_query_flags & (SD_RESOLVED_SYNTHETIC | SD_RESOLVED_FROM_TRUST_ANCHOR)) &&
```

Comment:
should probably be:

```c
return (q->answer_query_flags & SD_RESOLVED_SYNTHETIC) || (flags & SD_RESOLVED_FROM_MASK & ~(SD_RESOLVED_FROM_TRUST_ANCHOR|SD_RESOLVED_FROM_ZONE) == 0;
```

i.e. the synthetic bit is final: if its set, it's ours, and we should set AA. And on top of that anything from the trust anchor and from the host's mDNS/LLMNR zones are too

Reply:
But what if the answer includes both synthetic and external items? Then `SYNHETIC` will be set, but so will be `FROM_CACHE` (or something). And in that case we shouldn't claim to be authoritative for the whole answer, no?

Reply:
That being said, if you disallow multiple questions in one packet you're probably not going to end up with answers from different sources either, so maybe this is not a concern.

Reply:
answer are either synthetic or not. We don't merge synthetic with non-synthetic stuff in one DnsQuery. It's like being a "bit pregnant" ;-)

Reply:
Then this comment here sounds misleading:

https://github.com/systemd/systemd/blob/ff05157f828f2e5689bfbc101552566b623567f3/src/resolve/resolved-def.h#L58-L59

Should I remove `(at least partially)`? Or clarify it somehow?

Reply:
hmm, yeah, this is confusing.

so i added that comment because i wanted the fact that we do not partially synthesize to be an internal implementation detail, and that from the outside this should be opaque. these flags are after all exposed via D-Bus, i.e. they are public knowledge, and I wanted people not to rely on the fact that currently stuff is never half-synthesized, or never comes from more than one source. 

But maybe this was a stupid idea of mine, simply because for "half-synthesized" stuff, if we'd ever add that we probably should attach that info to the RRs indiviudally instead of the whole query anyway. And that means there's no benefit in claiming these bits are "partially" from some source or "partially" synthesized.

hence, please go ahead fix the comments, drop the `(at least partially)` everywhere, we should make it official that one answer is only from one source and not many, and only synthesized or not synthesized but never half-synthesized.

Reply:
> one answer is only from one source and not many, and only synthesized or not synthesized but never half-synthesized

In that case, this check can be simplified to just `(q->answer_query_flags & (SD_RESOLVED_SYNTHETIC|SD_RESOLVED_FROM_TRUST_ANCHOR|SD_RESOLVED_FROM_ZONE))`, right?

Reply:
yeah, i guess

Reply:
hmm, actually i might have talked nonsense above, i.e. there's the CNAME case, where we might merge data from multiple sources: the original CNAME record plus where it points to

Reply:
Right, that makes sense.

Then what happens (or what *should* happen) if you ask a remote server for `a.example.com` and get a
`a.example.com CNAME b.example.com`
and there's a static entry for `b.example.com` in `/etc/hosts`? Would/should resolved ignore the local entry in that case, or?..

Reply:
No AA bit in this case. It seems better to be conservative here.